### PR TITLE
Tie the M5Stack Temperature Cluster to the Virtual Thermometer

### PR DESCRIPTION
**Pull request by sagar-apple**
#### Problem
The M5Stack's virtual Temperature Sensor isn't actually hooked up the the Temperature Cluster it's running on Endpoint 1. 
We should tie them together so that users can edit the actual attribute's value from the M5Stack's UI.

 #### Summary of Changes
Tied the Virtual Temperature Attribute to the underlying Cluster running on the M5Stack.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: examples/all-clusters-app/esp32/main/component.mk [hunk 1]
```diff
@@ -44,6 +44,7 @@ COMPONENT_SRCDIRS :=
 
 COMPONENT_EXTRA_INCLUDES := $(PROJECT_PATH)/third_party/connectedhomeip/src/app/util                                   \
                             $(PROJECT_PATH)/third_party/connectedhomeip/src/app/server                                 \
+                            $(PROJECT_PATH)/third_party/connectedhomeip/src/app/clusters                               \
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/esp32/main/component.mk:47]: Why this change?

File: examples/all-clusters-app/esp32/main/main.cpp [hunk 2]
```diff
@@ -158,10 +160,18 @@ class EditAttributeListModel : public ListScreen::Model
...
         if (sscanf(value.c_str(), "%d", &n) == 1)
         {
+            auto & attribute = std::get<1>(std::get<1>(std::get<1>(devices[d])[e])[c])[a];
+            auto & name      = std::get<0>(attribute);
+
             ESP_LOGI(TAG, "editing attribute as integer: %d (%s)", n, i == 0 ? "+" : "-");
             n += (i == 0) ? 1 : -1;
             char buffer[32];
             sprintf(buffer, "%d", n);
+            if (!strncmp(name.c_str(), "Temperature", sizeof(name.c_str())))
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/esp32/main/main.cpp:170]: sizeof won't do the right thing here.  If `name` is an `std::string` here, why not just `if (name == "Temperature")`?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/all-clusters-app/esp32/main/component.mk (hunk 1 above)
Comment (sagar-apple) [reply examples/all-clusters-app/esp32/main/component.mk:47]: yeah looks like I didn't need this after all. Fixing.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@saurabhst @jelderton @andy31415 ?

**Review by msandstedt — APPROVED** [review]

File: examples/all-clusters-app/esp32/main/main.cpp [hunk 3]
```diff
@@ -301,9 +311,8 @@ void SetupPretendDevices()
     AddEndpoint("External");
     AddCluster("Thermometer");
     AddAttribute("Temperature", "21");
-    AddEndpoint("Internal");
-    AddCluster("Thermometer");
-    AddAttribute("Temperature", "42");
+    // write the temp attribute
+    emberAfPluginTemperatureMeasurementSetValueCallback(1, (int16_t)(21 * 100));
```
Comment (vivien-apple) [inline examples/all-clusters-app/esp32/main/main.cpp:315]: nit: `static_cast<int16_t>(21 * 100)`

File: src/app/clusters/temperature-measurement-server/temperature-measurement-server.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ */
+EmberAfStatus emberAfPluginTemperatureMeasurementSetValueCallback(chip::EndpointId endpoint, int16_t value);
+
+/** @brief Temperature Measurement Cluster Server Post Init
+ *
+ * Following resolution of the Temperature Measurement reading at startup for this endpoint,
+ * perform any additional initialization needed; e.g., synchronize hardware
+ * state.
+ */
+void emberAfPluginTemperatureMeasurementServerInitCallback(void);
```
Comment (vivien-apple) [inline src/app/clusters/temperature-measurement-server/temperature-measurement-server.h:38]: This is a bit weird but this one is already declared here: https://github.com/project-chip/connectedhomeip/blob/8c93ec6b67f9c37cfeec9fa1d2cb93266eb4aa95/src/app/util/util.cpp#L124

Comment (sagar-apple) [reply src/app/clusters/temperature-measurement-server/temperature-measurement-server.h:38]: Should I update the util file to consume this new header?

File: examples/all-clusters-app/esp32/main/main.cpp [hunk 5]
```diff
@@ -158,10 +160,18 @@ class EditAttributeListModel : public ListScreen::Model
         int n;
         if (sscanf(value.c_str(), "%d", &n) == 1)
         {
+            auto & attribute = std::get<1>(std::get<1>(std::get<1>(devices[d])[e])[c])[a];
```
Comment (andy31415) [inline examples/all-clusters-app/esp32/main/main.cpp:163]: Could we figure out a way to make this more readable?

**Review by andy31415 — APPROVED** [review]

File: src/app/clusters/temperature-measurement-server/temperature-measurement-server.h (hunk 4 above)
Comment (sagar-apple) [reply src/app/clusters/temperature-measurement-server/temperature-measurement-server.h]: Removed the definition in the header file, since Util seems to do its own weird thing here.

File: examples/all-clusters-app/esp32/main/main.cpp (hunk 5 above)
Comment (woody-apple) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: @sagar-apple ?

Comment (sagar-apple) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: Probably in a follow up PR no, @woody-apple?

Comment (woody-apple) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: Please file one, or fix. Ideally fix just some better variable names and splitting this out? Or event a comment... (I was cleaning out these todos in our issue tree, and... we have lots ;) )

Comment (sagar-apple) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: My assumption is that the single letter var names were chosen to keep this crazy long nested tuple access short. I'll rename and see what that does.

Comment (sagar-apple) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: Looking at it, it really think if this cleanup is needed it should be a separate PR. There's 4 nested classes all using single letter member vars.

Comment (andy31415) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: For sure a followup - this is used in other places.

Comment (sagar-apple) [reply examples/all-clusters-app/esp32/main/main.cpp:163]: I put this up https://github.com/project-chip/connectedhomeip/pull/4820, it's _slightly_ easier to read.
